diff -u3 -r tr7-v2.0.19.org/examples/tr7i.c tr7-v2.0.19/examples/tr7i.c
--- tr7-v2.0.19.org/examples/tr7i.c	2026-05-23 20:51:54.000000000 +0900
+++ tr7-v2.0.19/examples/tr7i.c	2026-08-30 08:36:08.514636100 +0900
@@ -10,6 +10,11 @@
 #ifdef _WIN32
 #  define snprintf _snprintf
 #endif
+#ifdef __MINGW32__
+#  define WIN32_LEAN_AND_MEAN
+#  include <windows.h>
+#  include <shlwapi.h>
+#endif
 
 #include "tr7.h"
 
@@ -83,6 +88,20 @@
    tr7_set_string(tsc, strid, val ? val : defval);
 }
 
+char *get_libpath(void) {
+   static char buf[1024] = {0};
+   static char relpath[] = "/../share/tr7/tr7libs";
+   uint32_t size = GetModuleFileNameA(NULL, buf, sizeof(buf));
+   if(size == sizeof(buf) || size == 0) return NULL;
+   PathRemoveFileSpecA(buf);
+   if (strlen(buf) + sizeof(relpath) > sizeof(buf)) {
+      fprintf(stderr, "library path is too long.\n");
+      exit(2);
+   }
+   strcat(buf, relpath);
+   return buf;
+}
+
 int main(int argc, char **argv)
 {
    tr7_engine_t tsc;
@@ -118,7 +137,11 @@
    tr7_set_standard_ports(tsc);
    tr7_set_argv(argv);
    addenvstr(tsc, Tr7_StrID_Path, PATHVAR, DEFPATH);
+#ifdef __MINGW32__
+   addenvstr(tsc, Tr7_StrID_Library_Path, LIBPATHVAR, get_libpath());
+#else
    addenvstr(tsc, Tr7_StrID_Library_Path, LIBPATHVAR, DEFLIBPATH);
+#endif
    addenvstr(tsc, Tr7_StrID_Include_Path, INCPATHVAR, DEFINCPATH);
    addenvstr(tsc, Tr7_StrID_Extension_Path, EXTPATHVAR, DEFEXTPATH);
    addenvstr(tsc, Tr7_StrID_Prompt, PROMPTVAR, PROMPT);
diff -u3 -r tr7-v2.0.19.org/Makefile tr7-v2.0.19/Makefile
--- tr7-v2.0.19.org/Makefile	2026-05-23 20:51:54.000000000 +0900
+++ tr7-v2.0.19/Makefile	2026-08-30 08:34:02.765007700 +0900
@@ -11,7 +11,7 @@
 .PHONY: all test chkopt clean updtref libtr7.so doc
 
 # tools
-CC ?= cc
+CC ?= gcc
 AR ?= ar
 RM ?= rm
 LN ?= ln
@@ -35,12 +35,12 @@
 FFLAGS =
 EFLAGS =
 MFLAGS = '-DVERSION="${VERSION}"' '-DMINVERSION="${MINVERSION}"'
-LDFLAGS = -Wl,--gc-sections,-z,relro,-z,now
+LDFLAGS = -Wl,--gc-sections
 # Detect darwin linker, which is not compatible with gnu linker
 ifneq ($(findstring PROJECT,$(shell $(LD) -v 2>&1)),)
 	LDFLAGS = -Wl,-dead_strip,-pie
 endif
-SYSLIBS += -lm
+SYSLIBS += -lshlwapi -lm
 
 # name of binary targets
 LIBSTA = libtr7.a
@@ -55,7 +55,7 @@
 INCDIR = $(PREFIX)/include
 DATADIR = $(PREFIX)/share
 TR7DIR = $(DATADIR)/tr7
-TR7LIBPATH = $(TR7DIR)/tr7libs:$(PWD)/tr7libs
+TR7LIBPATH = $(TR7DIR)/tr7libs
 
 # default targets
 ALL = $(LIBSTA) $(TR7I) tags doc
