diff --git a/lib/startup.wisp b/lib/startup.wisp
index 5ac3b1f..5fc1ac8 100755
--- a/lib/startup.wisp
+++ b/lib/startup.wisp
@@ -1,4 +1,4 @@
-(%define *require-path* '("lib" "."))
+(%define *require-path* (list (string-append (module-path) "/../share/wisp") "."))
 
 (%define *required* '())
 
diff --git a/src/convert.c b/src/convert.c
index 41a2dac..fd5e956 100644
--- a/src/convert.c
+++ b/src/convert.c
@@ -6,7 +6,9 @@
 #include <math.h>
 #include <stdlib.h>
 #include <string.h>
+#ifndef __MINGW32__
 #include <sys/errno.h>
+#endif
 
 enum {
     radix_unspecified = 0, // radix not known yet
diff --git a/src/core_types.h b/src/core_types.h
index 28e9982..b90da21 100644
--- a/src/core_types.h
+++ b/src/core_types.h
@@ -1,13 +1,21 @@
 #ifndef CORE_TYPES_H
 #define CORE_TYPES_H
 
+#ifdef __MINGW32__
+#include <windows.h>
+#include <shlwapi.h>
+#define WIN32_LEAN_AND_MEAN
+#endif
+
 #include <stdbool.h>
 #include <stdint.h>
 
+#ifndef __MINGW32__
 typedef bool BOOL;
 typedef char CHAR;
-typedef int64_t INT;
-typedef double FLOAT;
+#endif
+#define INT int64_t
+#define FLOAT double
 typedef uint16_t SLOT;
 typedef uint8_t TYPEID;
 
diff --git a/src/gc.c b/src/gc.c
index e3c2ab9..160d89c 100644
--- a/src/gc.c
+++ b/src/gc.c
@@ -13,7 +13,9 @@
 #include <math.h>
 #include <stdlib.h>
 #include <string.h>
+#ifndef __MINGW32__
 #include <sys/resource.h>
+#endif
 #include <sys/time.h>
 
 typedef struct {
@@ -82,6 +84,53 @@ static double mach_time_to_secs(uint64_t t) {
 }
 #endif
 
+#ifdef __MINGW32__
+# define RUSAGE_SELF 0
+# define RUSAGE_CHILDREN -1
+
+struct rusage {
+    struct timeval ru_utime;
+    struct timeval ru_stime;
+};
+
+int getrusage (int who, struct rusage *usage_p) {
+    if (who == RUSAGE_SELF || who == RUSAGE_CHILDREN) {
+        memset (usage_p, '\0', sizeof (struct rusage));
+
+        if (who == RUSAGE_SELF) {
+            FILETIME creation_time;
+            FILETIME exit_time;
+            FILETIME kernel_time;
+            FILETIME user_time;
+
+            if (GetProcessTimes (GetCurrentProcess (),
+                                                     &creation_time, &exit_time,
+                                                     &kernel_time, &user_time))
+                {
+                    uint64_t kernel_usec =
+                        ((((uint64_t) kernel_time.dwHighDateTime << 32)
+                            | (uint64_t) kernel_time.dwLowDateTime)
+                         + 5) / 10;
+                    uint64_t user_usec =
+                        ((((uint64_t) user_time.dwHighDateTime << 32)
+                            | (uint64_t) user_time.dwLowDateTime)
+                         + 5) / 10;
+
+                    usage_p->ru_utime.tv_sec = user_usec / 1000000U;
+                    usage_p->ru_utime.tv_usec = user_usec % 1000000U;
+                    usage_p->ru_stime.tv_sec = kernel_usec / 1000000U;
+                    usage_p->ru_stime.tv_usec = kernel_usec % 1000000U;
+                }
+        }
+        return 0;
+    }
+    else {
+        errno = EINVAL;
+        return -1;
+    }
+}
+#endif
+
 typedef struct {
     INT num_collections;
     size_t bytes_total;
diff --git a/src/main.c b/src/main.c
index b270780..971c064 100644
--- a/src/main.c
+++ b/src/main.c
@@ -59,7 +59,54 @@ CELL register_func(const FUNC_META *meta) {
     return result;
 }
 
+#ifdef __MINGW32__
+const char *internal_module_path(void) {
+    static char buf[1024] = "";
+    uint32_t size = GetModuleFileNameA(NULL, buf, sizeof(buf));
+    if(size == sizeof(buf) || size == 0) return NULL;
+    PathRemoveFileSpecA(buf);
+    return buf;
+}
+
+char *make_startup_path(void) {
+    const char *module_path = internal_module_path();
+    if(module_path == NULL) return NULL;
+    static const char rel[] = "/../share/wisp/startup.wisp";
+    size_t len = strlen(module_path)+sizeof(rel)+1;
+    char *newstr = malloc(len);
+    strcpy(newstr, module_path);
+    strcat(newstr, rel);
+    return newstr;
+}
+
+DECLARE_FUNC_0(func_module_path, "module-path", "Returns module path as string.")
+
+CELL func_module_path(CELL frame) {
+    const char *path = internal_module_path();
+    if(path == NULL) return make_exception("unable to get module filename.");
+    CELL module_path = make_string(path);
+    return module_path;
+}
+
+DECLARE_FUNC(
+    func_path_relative, 1, 1,
+    "path-relative?", "string",
+    "Returns #t if <string> is rerative path, otherwise #f."
+)
+
+CELL func_path_relative(CELL frame) {
+    if (!STRINGP(FV0)) {
+        return make_exception("expects string");
+    }
+    return MKBOOL(PathIsRelativeA(GET_STRING(FV0)->data));
+}
+#endif
+
 void core_register_symbols() {
+#ifdef __MINGW32__
+    register_func(&meta_func_module_path);
+    register_func(&meta_func_path_relative);
+#endif
 }
 
 extern void arith_register_symbols();
@@ -336,7 +383,18 @@ int main(int argc, char *argv[]) {
     register_symbols();
 
     if (opt_load_libs) {
+#ifdef __MINGW32__
+        char *startup_path = make_startup_path();
+        if(startup_path == NULL) {
+            printf("unable to get module filename.");
+            free(startup_path);
+            exit(1);
+        }
+        const CELL result = internal_load(startup_path);
+        free(startup_path);
+#else
         const CELL result = internal_load("lib/startup.wisp");
+#endif
         if (EXCEPTIONP(result)) {
             print_result(result);
             exit(1);
diff --git a/src/read.c b/src/read.c
index 05fd1dd..5247456 100644
--- a/src/read.c
+++ b/src/read.c
@@ -91,6 +91,61 @@ const char *internal_get_prompt(char *buf, size_t cap) {
     return buf;
 }
 
+#ifdef __MINGW32__
+ssize_t getline(char **lineptr, size_t *n, FILE *stream) {
+    char *buf, *ptr;
+    size_t size;
+    int c;
+    int delimiter = '\n';
+
+    if (!lineptr || !n || !stream) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    buf = *lineptr;
+    size = *n;
+
+    if (buf == NULL || size == 0) {
+        size = 128;
+        buf = malloc(size);
+        if (buf == NULL) {
+            errno = ENOMEM;
+            return -1;
+        }
+    }
+
+    ptr = buf;
+
+    while ((c = getc(stream)) != EOF) {
+        if ((size_t)(ptr - buf) + 1 >= size) {
+            size_t new_size = size * 2;
+            char *new_buf = realloc(buf, new_size);
+            if (!new_buf) {
+                errno = ENOMEM;
+                return -1;
+            }
+            ptr = new_buf + (ptr - buf);
+            buf = new_buf;
+            size = new_size;
+        }
+
+        *ptr++ = c;
+        if (c == delimiter)
+            break;
+    }
+
+    if (ptr == buf && c == EOF)
+        return -1;
+
+    *ptr = '\0';
+    *lineptr = buf;
+    *n = size;
+
+    return ptr - buf;
+}
+#endif
+
 CELL internal_read_line(FILE *fp) {
 #ifdef HAVE_READLINE
     if (fp == stdin) {
diff --git a/src/wisp.h b/src/wisp.h
index 02be120..2681379 100644
--- a/src/wisp.h
+++ b/src/wisp.h
@@ -658,6 +658,7 @@ static inline CELL make_slot(SLOT slot) {
 static inline CELL make_bool(bool b) {
     return b ? V_TRUE : V_FALSE;
 }
+#define MKBOOL make_bool
 
 static inline CELL make_indirect_tag(TYPEID type) {
     CELL cell = {.as_bits = (uint64_t) T_INDIRECT_TAG << 45 | (uint64_t) type};
