diff --git a/src/foment.hpp b/src/foment.hpp
index 745d5e4..42a29c3 100644
--- a/src/foment.hpp
+++ b/src/foment.hpp
@@ -81,7 +81,11 @@ typedef double double64_t;
 typedef wchar_t FCh16;
 typedef FCh16 FChS;
 
+#ifdef __MINGW32__
+#define FALIGN __attribute__ ((aligned(8)))
+#else
 #define FALIGN __declspec(align(8))
+#endif
 
 #ifdef _M_AMD64
 #define FOMENT_64BIT
diff --git a/src/io.cpp b/src/io.cpp
index 329f51c..3143a66 100644
--- a/src/io.cpp
+++ b/src/io.cpp
@@ -4299,7 +4299,7 @@ Define("get-ip-addresses", GetIpAddressesPrimitive)(long_t argc, FObject argv[])
                 {
                     char buf[46];
 
-                    lst = MakePair(MakeStringC(InetNtop(AF_INET,
+                    lst = MakePair(MakeStringC(inet_ntop(AF_INET,
                             &((struct sockaddr_in *) iaua->Address.lpSockaddr)->sin_addr, buf, 46)),
                             lst);
                 }
@@ -4307,7 +4307,7 @@ Define("get-ip-addresses", GetIpAddressesPrimitive)(long_t argc, FObject argv[])
                 {
                     char buf[46];
 
-                    lst = MakePair(MakeStringC(InetNtop(AF_INET6,
+                    lst = MakePair(MakeStringC(inet_ntop(AF_INET6,
                             &((struct sockaddr_in6 *) iaua->Address.lpSockaddr)->sin6_addr, buf,
                             46)), lst);
                 }
diff --git a/src/txt2cpp.cpp b/src/txt2cpp.cpp
index 536290b..636db3c 100644
--- a/src/txt2cpp.cpp
+++ b/src/txt2cpp.cpp
@@ -4,7 +4,7 @@ Convert a scheme library into a multiple line string constant.
 
 */
 
-#ifdef FOMENT_WINDOWS
+#if defined(FOMENT_WINDOWS) && !(__MINGW32__)
 #define _CRT_SECURE_NO_WARNINGS
 #define strdup(s) _strdup(s)
 #endif // FOMENT_WINDOWS
